bug: panic in MultiAgent.UnsubscribeAgent with ExperimentSingleTailnet

Panic under rPTY traffic with single_tailnet enabled.

https://stats.dev.c8s.io/d/qLVSTR-Vz/coderv2-loadtest-dashboard?orgId=1&from=1695829564765&to=1695829732027

Logs: https://cloudlogging.app.goo.gl/bTGg3wyNM8iLf6nY7

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x137db65]

goroutine 286 [running]:
github.com/coder/coder/v2/tailnet.(*MultiAgent).UnsubscribeAgent(0xfd7a115ca1e04c18?, {0xa, 0xd1, 0x65, 0xf4, 0x9f, 0x48, 0x4c, 0x18, 0xac, ...})
	/home/runner/actions-runner/_work/coder/coder/tailnet/multiagent.go:104 +0xc5
github.com/coder/coder/v2/coderd.(*ServerTailnet).doExpireOldAgents(0xc0001af040, 0x1a3185c5000)
	/home/runner/actions-runner/_work/coder/coder/coderd/tailnet.go:205 +0x72f
github.com/coder/coder/v2/coderd.(*ServerTailnet).expireOldAgents(0xc0001af040)
	/home/runner/actions-runner/_work/coder/coder/coderd/tailnet.go:172 +0x92
created by github.com/coder/coder/v2/coderd.NewServerTailnet
	/home/runner/actions-runner/_work/coder/coder/coderd/tailnet.go:152 **+0xc9f**
```

--- a/enterprise/tailnet/coordinator.go
+++ b/enterprise/tailnet/coordinator.go
@@ -57,8 +57,9 @@ func (c *haCoordinator) ServeMultiAgent(id uuid.UUID) agpl.MultiAgentConn {
 		ID:                id,
 		AgentIsLegacyFunc: c.agentIsLegacy,
 		OnSubscribe:       c.clientSubscribeToAgent,
+		OnUnsubscribe:     c.clientUnsubscribeFromAgent,
 		OnNodeUpdate:      c.clientNodeUpdate,
-		OnRemove:          func(enq agpl.Queue) { c.clientDisconnected(enq.UniqueID()) },
+		OnRemove:          c.clientDisconnected,
 	}).Init()
 	c.addClient(id, m)
 	return m
@@ -99,6 +100,22 @@ func (c *haCoordinator) clientSubscribeToAgent(enq agpl.Queue, agentID uuid.UUID
 
 	// nolint:nilnil
 	return nil, nil
+}
+
+func (c *haCoordinator) clientUnsubscribeFromAgent(enq agpl.Queue, agentID uuid.UUID) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	connectionSockets, ok := c.agentToConnectionSockets[agentID]
+	if !ok {
+		return nil
+	}
+	delete(connectionSockets, enq.UniqueID())
+	if len(connectionSockets) == 0 {
+		delete(c.agentToConnectionSockets, agentID)
+	}
+
+	return nil
 }
 
 type haCoordinator struct {
@@ -161,7 +178,7 @@ func (c *haCoordinator) ServeClient(conn net.Conn, id, agentID uuid.UUID) error 
 	defer tc.Close()
 
 	c.addClient(id, tc)
-	defer c.clientDisconnected(id)
+	defer c.clientDisconnected(tc)
 
 	agentNode, err := c.clientSubscribeToAgent(tc, agentID)
 	if err != nil {
@@ -200,26 +217,24 @@ func (c *haCoordinator) initOrSetAgentConnectionSocketLocked(agentID uuid.UUID, 
 	c.clientsToAgents[enq.UniqueID()][agentID] = c.agentSockets[agentID]
 }
 
-func (c *haCoordinator) clientDisconnected(id uuid.UUID) {
+func (c *haCoordinator) clientDisconnected(enq agpl.Queue) {
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
 
-	for agentID := range c.clientsToAgents[id] {
-		// Clean all traces of this connection from the map.
-		delete(c.nodes, id)
+	for agentID := range c.clientsToAgents[enq.UniqueID()] {
 		connectionSockets, ok := c.agentToConnectionSockets[agentID]
 		if !ok {
-			return
+			continue
 		}
-		delete(connectionSockets, id)
-		if len(connectionSockets) != 0 {
-			return
+		delete(connectionSockets, enq.UniqueID())
+		if len(connectionSockets) == 0 {
+			delete(c.agentToConnectionSockets, agentID)
 		}
-		delete(c.agentToConnectionSockets, agentID)
 	}
 
-	delete(c.clients, id)
-	delete(c.clientsToAgents, id)
+	delete(c.nodes, enq.UniqueID())
+	delete(c.clients, enq.UniqueID())
+	delete(c.clientsToAgents, enq.UniqueID())
 }
 
 func (c *haCoordinator) handleNextClientMessage(id uuid.UUID, decoder *json.Decoder) error {

